packages/backend/ubuntu-pro_test.go

```diff
@@ -0,0 +1,460 @@
+package main
+
+import (
+	"context"
+	"testing"
+	//epb "google.golang.org/protobuf/types/known/emptypb"
+	"fmt"
+	pb "github.com/canonical/desktop-security-center/packages/proto"
+	"github.com/canonical/desktop-security-center/packages/testutils"
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	"os"
+	"os/exec"
+)
+
+type prodbus struct{}
+const (
+    waitJson = `{
+              "_schema_version": "v1",
+              "data": {
+                "attributes": {
+                  "contract_token": "%s"
+                },
+                "meta": {
+                  "environment_vars": []
+                },
+                "type": "MagicAttachInitiate"
+              },
+              "errors": [%s],
+              "result": "%s",
+              "version": "30.1",
+              "warnings": []
+            }`
+    initiateJson = `{
+              "_schema_version": "v1",
+              "data": {
+                "attributes": {
+                  "expires_in": %d,
+                  "token": "%s",
+                  "user_code": "%s"
+                },
+                "meta": {
+                  "environment_vars": []
+                },
+                "type": "MagicAttachInitiate"
+              },
+              "errors": [%s],
+              "result": "%s",
+              "version": "30.1",
+              "warnings": []
+            }`
+)
+
+/* Variables for adjusting test results */
+var getResponse string
+var failEnable bool
+var failDisable bool
+
+var manager *Manager
+var ctx context.Context
+
+func TestWaitProMagicFlow(t *testing.T) {
+    cases := []struct {
+        name string
+        errors string
+        contract string
+        result string
+        wantError bool
+    }{
+        {
+            name: "Confirmed contract",
+            contract: "Ca1xQ10aB9Xpp",
+            result: "success",
+        },
+        {
+            name: "Invalid response, broken JSON",
+            wantError: true,
+            contract: ",\"\"{{{",
+            result: "success",
+        },
+        {
+            name: "Invalid response, has failure",
+            wantError: true,
+            result: "failure",
+            errors: `{
+                        "msg": "Failed to connect to authentication server",
+                        "code": "connectivity-error",
+                        "meta": {}
+                     }`,
+        },
+    }
+    for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+            execMocker := func(name string, arg ...string) *exec.Cmd {
+                ret := fmt.Sprintf(waitJson, tc.contract, tc.errors, tc.result)
+                return exec.Command("printf", "%s", ret)
+            }
```

Comment:
Nice work mocking the exec! What would make this a little cleaner would be to:

- Create the `execMocker` as a type
- Have boolean flags you could set depending on what error you want it to return (ie: connectivity-error)
- Create the mock with the flags and move the logic of what is returned into the mock, not passing it in the test

The main goal would be to make your test cases as concise as possible, so instead of:
```

           name: "Invalid response, has failure",
            wantError: true,
            result: "failure",
            errors: `{
                        "msg": "Failed to connect to authentication server",
                        "code": "connectivity-error",
                        "meta": {}
                     }`,
```

you could have:
```
            name: "Invalid response, has failure",
            wantError: true,
            networkError: true
```

Reply:
Something like this in the last commit?

Reply:
I'm reverting that last commit because I want to merge this but not with a "surprise" commit. I will include it in the other merge request with the existing Github actions you mentioned in another comment.